src/main/java/week3/WordSearch.java

```diff
@@ -0,0 +1,195 @@
+package week3;
+
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class WordSearch {
+
+  /*
+   * Given a grid of letters and a dictionary, find all the words from the dictionary that can be
+   * formed in the grid.The rules for forming a word:
+   *    ● You can start at any position.
+   *    ● You can move to one of the 8 adjacent cells (horizontally/vertically/diagonally).
+   *    ● You can't visit the same cell twice in the same word.Assume you have a dictionary class
+   *      with these two methods:
+   *    ● isWord(string): Returns whether the given string is a valid word.
+   *    ● isPrefix(string): Returns whether the given string is a prefix of at least one word in
+   *      the dictionary.
+   *
+   * Your function receives the number of rows, number of columns, a 2-dimensional array of
+   * characters (of the native char data type), and the dictionary. You should return the set of all
+   * words found.
+   */
+
+  private int rows;
+  private int cols;
```

Comment:
I don't like these variables. They represent the state of the currently-running algorithm, rather than the object.

They prevent the algorithm being used reentrantly - you can't call findWords twice simultaneously on different threads. They make the flow of data less obvious - you have to do a search to see where they were assigned to and used.

I would recommend either: a) Creating an object specifically for each run. It would have member variables holding the grid, grid size, dictionary, etc. b) Pass the extra parameters explicitly as arguments to the method.